src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -0,0 +1,228 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    /// <summary>
+    ///     Applies <see cref="IProjectVersionedValue{T}"/> values to a <see cref="IWorkspaceProjectContext"/>.
+    /// </summary>
+    [Export(typeof(IApplyChangesToWorkspaceContext))]
+    internal class ApplyChangesToWorkspaceContext : OnceInitializedOnceDisposed, IApplyChangesToWorkspaceContext
+    {
+        private const string DesignTimeRuleName = CompilerCommandLineArgs.SchemaName;
+        private readonly ConfiguredProject _project;
+        private readonly ICommandLineParserService _commandLineParser;
+        private readonly IProjectLogger _logger;
+
+        private IWorkspaceProjectContext _context;
+        private ExportLifetimeContext<IWorkspaceContextHandler>[] _handlers;
+
+        [ImportingConstructor]
+        public ApplyChangesToWorkspaceContext(ConfiguredProject project, ICommandLineParserService commandLineParser, IProjectLogger logger)
+        {
+            _project = project;
+            _commandLineParser = commandLineParser;
+            _logger = logger;
+
+            WorkspaceContextHandlerFactories = new OrderPrecedenceExportFactoryCollection<IWorkspaceContextHandler>();
+        }
+
+        [ImportMany]
+        public OrderPrecedenceExportFactoryCollection<IWorkspaceContextHandler> WorkspaceContextHandlerFactories
+        {
+            get;
+        }
+
+        public void Initialize(IWorkspaceProjectContext context)
+        {
+            Requires.NotNull(context, nameof(context));
+
+            lock (SyncObject)
+            {
+                if (_context != null)
+                    throw new InvalidOperationException("Already initialized.");
+
+                _context = context;
+
+                EnsureInitialized();
+            }
+        }
+
+        public void ApplyDesignTime(IProjectVersionedValue<IProjectSubscriptionUpdate> update, bool isActiveContext, CancellationToken cancellationToken)
+        {
+            Requires.NotNull(update, nameof(update));
+            
+            lock (SyncObject)
+            {
+                VerifyInitializedAndNotDisposed();
+
+                IProjectChangeDescription projectChange = update.Value.ProjectChanges[DesignTimeRuleName];
+
+                if (projectChange.Difference.AnyChanges)
+                {
+                    IComparable version = GetConfiguredProjectVersion(update);
+
+                    ProcessOptions(projectChange.After);
+                    ProcessCommandLine(version, projectChange.Difference, isActiveContext, cancellationToken);
+                    ProcessDesignTimeBuildFailure(projectChange.After);
+                }
+            }
+        }
+
+        public void ApplyEvaluation(IProjectVersionedValue<IProjectSubscriptionUpdate> update, bool isActiveContext, CancellationToken cancellationToken)
+        {
+            Requires.NotNull(update, nameof(update));
+
+            lock (SyncObject)
```

Comment:
We lock to avoid overlapping ApplyDesignTime/ApplyEvaluation and have both touch IWorkspaceProjectContext at the same time. Note, this will currently block a thread waiting for this when they overlap - which might be problematic. I'm looking at coordinating this in a non-blocking mechanism I start testing this over large code bases.